File: backend/core/src/portfolio/pnl_calculator.cpp
#include "portfolio_state.hpp"

#include <cmath>

namespace sovereign {

PortfolioMetrics PnlCalculator::calculate(const PortfolioState& state) {
    PortfolioMetrics metrics;
    metrics.total_equity = state.cash;
    metrics.total_unrealized_pnl = 0.0;
    metrics.net_exposure = 0.0;
    metrics.gross_exposure = 0.0;

    for (const auto& pos : state.positions) {
        const double market_value = pos.quantity * pos.current_price;
        const double cost_basis = pos.quantity * pos.average_cost;
        const double unrealized_pnl = market_value - cost_basis;

        metrics.total_equity += market_value;
        metrics.total_unrealized_pnl += unrealized_pnl;
        metrics.net_exposure += market_value;
        metrics.gross_exposure += std::abs(market_value);
    }

    if (metrics.total_equity > 0.0) {
        metrics.total_exposure = metrics.gross_exposure / metrics.total_equity;
    } else {
        metrics.total_exposure = 0.0;
    }

    metrics.ok = true;
    return metrics;
}

} // namespace sovereign

File: backend/cli/src/portfolio/pnl_calculator.rs
use super::state::PortfolioState;

#[derive(Debug, Clone, Default)]
pub struct PortfolioMetrics {
    pub total_equity: f64,
    pub unrealized_pnl: f64,
    pub net_exposure: f64,
    pub gross_exposure: f64,
}

pub fn calculate(state: &PortfolioState) -> PortfolioMetrics {
    let mut metrics = PortfolioMetrics {
        total_equity: state.cash,
        ..PortfolioMetrics::default()
    };

    for position in &state.positions {
        let market_value = position.quantity * position.current_price;
        let cost_basis = position.quantity * position.average_cost;
        metrics.total_equity += market_value;
        metrics.unrealized_pnl += market_value - cost_basis;
        metrics.net_exposure += market_value;
        metrics.gross_exposure += market_value.abs();
    }

    metrics
}
